exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java

```diff
@@ -47,6 +49,14 @@
   public List<String> files;
   public final String selectionRoot;
 
+  private enum StatusType {
+    CHECKED_DIRS,        // whether we have already checked for directories
+    HAS_DIRS,            // whether directories were found in the selection
+    EXPANDED_DIRS        // whether this selection has been expanded to files
+  }
+
+  private final BitSet dirStatus;
```

Comment:
You took this differently than I meant it. My proposal was that FileSelection has various states:

NOT_CHECKED_DIRS => (HAS_DIRS | NO_DIRS) => EXPANDED

Doesn this lifecycle describe the state of FileSelection? This way you don't have the multi-state-management problem you currently have below with this kind of construct: 

    fileSel.setExpanded(true);
    fileSel.setCheckedForDirectories(true);
    fileSel.setHasDirectories(false);  // already expanded

For each of the enumerations, we can return the right booleans that you need through enumerator constructors.